### PR TITLE
[IMP] mail: display edited at the end of the message

### PR DESCRIPTION
**Pull request by ibo-odoo**
**PURPOSE**

Display to other users the information that a message has been edited.

**SPECIFICATION**

If the user edits a message, (edited) in muted is displayed at the end of the
message and hovering on (edited) should indicate the date and time at which
the message was last modified.

**Task**-2664848

### CODE REVIEW

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -85,6 +85,7 @@ def default_get(self, fields):
     # content
     subject = fields.Char('Subject')
     date = fields.Datetime('Date', default=fields.Datetime.now)
+    edit_date = fields.Datetime('last edited DateTime', readonly=1)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:88]: To keep model as simple as possible, you may use write_date instead. As messages are never edited this should be sufficient in most use cases and an unnecessary column on an heavily used model.

Comment (ibo-odoo) [reply addons/mail/models/mail_message.py:88]: As discussed earlier, In the case of starred message write_date is also updated. So we won't get the proper write_date in the case to put the `edited` at the end of the message.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_message.py:88]: We have several millions of mail.message in DB, adding a column when 95% of messages are not edited is globally wasting space. Editing message is something you globally don't do except the discuss use case. I still think write_date is sufficient.

The other way round would be to store extra information on another table (edit_date, star status, ...). That way we lessen noise on a big table.

File: addons/mail/static/src/components/message/message.js [hunk 2]
```diff
@@ -116,6 +116,15 @@ export class Message extends Component {
...
 
+    /**
+     * Get the date time of the edited message at current user locale time.
+     *
+     * @returns {string}
+     */
+    get editdatetime() {
+        return this.messageView.message.editDate.format(getLangDatetimeFormat());
+    }
+
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.js:127]: No getter in component, replace this by introducing field `editDatetime` on JS model `Message`.

Comment (ibo-odoo) [reply addons/mail/static/src/components/message/message.js]: done.

**Comment by tde-banana-odoo** [comment]
Hello @ibo-odoo and @alexkuhn ,

Pointing out again you are adding model complexity for a feature used in few use case (think of email usage of chat notably, few people really use discuss and backend when not using the chat app). Cleaning mail.message model has been done between Odoo 13 and 15 to remove most of complexity and try to have a model as compact as possible. It contains fields that should not exist because discuss still does not use its own messaging model, which was planned for 15.0 but not done. Until then I understand some fields have to exist but it seems we have alternatives here.

I see at least two possible alternatives
  * use ``write_date``. @ibo-odoo pointed out that it is used for starring messages but it seems easy to me to skip the write on mail.message and instead write directly on the link table if necessary;
  * use another table, with maybe other fields to move on that table that would be discuss related;

It could be great to find a clean solution for that issue, as it is a real issue. Could you check more in details ?

Cheers,

**Comment by alexkuhn** [comment]
> I see at least two possible alternatives
> 
> * use `write_date`. @ibo-odoo pointed out that it is used for starring messages but it seems easy to me to skip the write on mail.message and instead write directly on the link table if necessary;

Don't like this solution: write date is too technical and not necessarily the same as edit date, so it's too fragile to make this shortcut.

> * use another table, with maybe other fields to move on that table that would be discuss related;

Good idea. The plan is for discuss to use its own messaging models, so it's a step in the right direction.

cc @seb-odoo

**Comment by seb-odoo** [comment]
`write_date` looks like a waste of space on the message table (I don't know for the rest of the code with certainty, but we never  use it).

We could probably remove the auto `write_date` field that has no clear purpose (and probably also remove `write_uid`), and introduce `edit_date` instead, that is filled only on edit.

While we are at it `create_uid` looks useless as there is already `author_id`, and `create_date` I kind of get it's different from `date` for incoming emails, but in discuss the difference between the two is also completely useless.

That said if we want to create an "edit history" table with each date & editor like github does for their comments to not have any field at all in message for that, I would be fine with it too.

**Comment by tde-banana-odoo** [comment]
@seb-odoo Short version would be to use ``write_date`` even if not technically always correct due to star (even if I still think this can be avoided by tweaking the star update). But doing the full version by moving all that in a separate table + maybe checking if we can remove the auto fields would be better from a model cleaning point of view.

It is up to you I would say. If this is not top-notch-urgent I would vote for the cleaning model version.

**Comment by seb-odoo** [comment]
It's not urgent, the current priority is perf not feature, so actually was is more "urgent" is to do the cleaning of the table and remove the auto-fields if possible.

**Comment by seb-odoo** [comment]
This was done in another PR

**Pull request closed without merging**